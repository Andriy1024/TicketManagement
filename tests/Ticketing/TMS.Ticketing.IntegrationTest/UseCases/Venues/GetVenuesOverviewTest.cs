using TMS.Ticketing.Application.Dtos;
using TMS.Ticketing.Application.Repositories;
using TMS.Ticketing.Application.Services.Payments;
using TMS.Ticketing.Application.UseCases.Venues;
using TMS.Ticketing.IntegrationTest.Common;
using TMS.Ticketing.IntegrationTest.Common.Factories;
using TMS.Ticketing.IntegrationTest.UseCases.Venues.Common;

namespace TMS.Ticketing.IntegrationTest.UseCases.Venues;

[Collection(VenuesDatabaseCollection.Name)]
public class GetVenuesOverviewTest
{
    private readonly TicketingServices _services;

    public GetVenuesOverviewTest(MongoDbFactory mongoDb)
    {
        _services = new TicketingServices()
          .AddJsonConfig("appsettings", "appsettings.test.json")
          .AddMongoConnection(mongoDb.ConnectionString, Guid.NewGuid().ToString())
          .BuildConfiguration()
          .AddTicketingServices()
          .OverrideService(Substitute.For<IPaymentsService>());
    }

    [Fact]
    public async Task GetVenuesOverview_ByDefaultReturns_VenueOverviewDtos()
    {
        // Arrange
        var services = _services.BuildServiceProvider();

        await SeedDataAsync(services);

        IEnumerable<VenueOverviewDto> actual;

        // Act
        using (var scope = services.CreateScope()) 
        {
            var handler = scope.ServiceProvider.GetRequiredService<IRequestHandler<GetVenuesOverview, IEnumerable<VenueOverviewDto>>>();

            actual = await handler.Handle(new GetVenuesOverview(), CancellationToken.None);
        }

        // Assert
        actual.Should().NotBeNull().And.HaveCount(2);

        actual.Should().MatchSnapshot(nameof(actual), matchOptions: o => 
        {
            o.IgnoreField($"**.{nameof(VenueOverviewDto.Id)}");

            return o;
        });
    }

    private static async Task SeedDataAsync(IServiceProvider services) 
    {
        using var scope = services.CreateScope();

        var repo = scope.ServiceProvider.GetRequiredService<IVenuesRepository>();

        var firstVenue = VenueTestFactory.Create(name: "Venue #1");
        var secondVenue = VenueTestFactory.Create(name: "Venue #2");

        await repo.AddAsync(firstVenue);
        await repo.AddAsync(secondVenue);
    }
}