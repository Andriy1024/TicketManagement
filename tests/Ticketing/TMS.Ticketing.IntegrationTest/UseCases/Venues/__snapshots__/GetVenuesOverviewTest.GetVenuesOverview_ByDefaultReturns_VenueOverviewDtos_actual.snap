[
  {
    "Id": "0e7d6f7f-ec5d-4c1e-a707-5e7be6bbb837",
    "Name": "Venue #1",
    "Country": "USA",
    "City": "Denver",
    "Street": "Steet 45",
    "Details": [
      {
        "Name": "Detail 1",
        "Value": "Detail Value"
      }
    ]
  },
  {
    "Id": "be85e499-5f9b-48b7-8dee-724ccebdedab",
    "Name": "Venue #2",
    "Country": "USA",
    "City": "Denver",
    "Street": "Steet 45",
    "Details": [
      {
        "Name": "Detail 1",
        "Value": "Detail Value"
      }
    ]
  }
]
